### ISSUE TITLE
Creating variables in `match` (`PHP 8`) can lead to `PossiblyUndefinedVariable` and `MixedArgument` false-positive warnings

### ISSUE THREAD

**Issue by AlexMinaev19** [vimeo/psalm#8320]
If you will try to create a variable in a new `match` expression you can get  `PossiblyUndefinedVariable` and `MixedArgument` false-positive warnings.

I did some examples: https://psalm.dev/r/3fcba079df

With the `switch` statement it works fine and non-false-positive warnings are not appearing.

**Comment by AndrolGenhald** [comment]
In Rust and Scala (where `match` is taken from), assigning the value of the match expression itself is more idiomatic anyway:

```php
$object = match($case) {
    'a' => new A(),
    'b' => new B(),
    default => throw new RuntimeException("Undefined case: $case")
};
```

https://psalm.dev/r/e01ceb02bf
